### PR TITLE
Fix parallel dispatch with target argument in transpile()

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

If a custom target was specified when calling the transpile() function with more than one circuit the parallel dispatch of the shared custom target would fail because the basis_gates created from the target was a dict keys type. Pickle is unable to serialize a dict keys object apparently and this would cause an exception to be raised. This commit fixes this by casting the basis_gates from the target as a list so that it can be serialized.

### CODE REVIEW

File: test/python/compiler/test_transpiler.py [hunk 1]
```diff
@@ -1617,3 +1617,31 @@ def test_custom_multiple_circuits(self):
...
+
+    def setUp(self):
+        super().setUp()
+        original_value = os.getenv("QISKIT_PARALLEL", None)
+
+        def restore_parallel():
+            os.environ["QISKIT_PARALLEL"] = original_value
+
+        self.addCleanup(restore_parallel)
+        del os.environ["QISKIT_PARALLEL"]
```
Comment (jakelishman) [inline test/python/compiler/test_transpiler.py:1634]: This implicitly assumes that `QISKIT_PARALLEL` is in the environment, which might not be the case for people calling `stestr` directly.  We only need the cleanup and the `del` if `original_value is not None` on the first call, I think.

Comment (kevinhartman) [reply test/python/compiler/test_transpiler.py:1634]: I wonder if we could just use `unittest.mock.patch` as a class decorator to temporarily update `os.environ`?

https://docs.python.org/3/library/unittest.mock.html#patch-dict

Comment (mtreinish) [reply test/python/compiler/test_transpiler.py:1634]: `patch.dict` is a good use case for this, I'll modify it now

Comment (mtreinish) [reply test/python/compiler/test_transpiler.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/8952/commits/949703cf31b9133f211ba612626b7bd0c942d399

**Review by jakelishman — APPROVED** [review]
Looks good to me - it's good to get all these parallel issues squashed.

**Comment by mtreinish** [comment]
I'm thinking now that we have a pattern for running things in parallel in the unit tests that we can drop the standalone script we run after the unittests and integrate that into the test suite. I'll do that in a follow up PR

**Review by kevinhartman — DISMISSED** [review]

File: test/python/compiler/test_transpiler.py [hunk 2]
```diff
@@ -1617,3 +1617,22 @@ def test_custom_multiple_circuits(self):
...
+
+    @data(0, 1, 2, 3)
+    def test_parallel_with_target(self, opt_level):
+        """Test that parallel dispatch works with a manual target."""
+        qc = QuantumCircuit(2)
+        qc.h(0)
+        qc.cx(0, 1)
+        qc.measure_all()
+        target = FakeMumbaiV2().target
+        with patch.dict("os.environ", {"QISKIT_PARALLEL": "FALSE"}):
```
Comment (mtreinish) [inline test/python/compiler/test_transpiler.py:1634]: Oops, this is backwards:

```suggestion
        with patch.dict("os.environ", {"QISKIT_PARALLEL": "TRUE"}):
```

**Review by kevinhartman — APPROVED** [review]
LGTM!

**Comment by mtreinish** [comment]
Playing with the tests more locally I don't think the mock is working as expected. I'm not able to trigger the bug with the tests without the code change applied and adding a print in `parallel_map()` isn't being displayed when running the tests. I've marked this as on hold to block merging until I figure out the mock pattern

**Comment by mtreinish** [comment]
ok https://github.com/Qiskit/qiskit-terra/pull/8952/commits/9d2cfd2a67ba8a8f85c3d784cee92b32b213e760 should fix it, it turns out the env variable approach would never have worked because it's only checked once on import.

**Pull request merged**